### PR TITLE
Move Tenant- and TimelineInfo structs to models.rs.

### PR DESCRIPTION
**Pull request by hlinnaka**
They are part of the management API response structs. Let's try to
concentrate everything that's part of the API in models.rs.

### CODE REVIEW

File: pageserver/src/timelines.rs [hunk 1]
```diff
@@ -33,123 +32,80 @@ use crate::{layered_repository::LayeredRepository, walredo::WalRedoManager};
...
+pub fn get_info_from_repo_timeline<T>(
+    tenant_id: ZTenantId,
+    timeline_id: ZTimelineId,
+    repo_timeline: &RepositoryTimeline<T>,
+    include_non_incremental_logical_size: bool,
+    include_non_incremental_physical_size: bool,
+) -> anyhow::Result<LocalTimelineInfo> {
+    match repo_timeline {
+        RepositoryTimeline::Loaded(_) => {
+            let timeline = tenant_mgr::get_local_timeline_with_load(tenant_id, timeline_id)?;
```
Comment (LizardWizzard) [inline pageserver/src/timelines.rs:100]: Now we can avoid that. This is more of an oversight from DatadirTimeline PR. Now we can pass `&RepositoryTimeline<TimelineImpl>` and use timeline from `Loaded` variant directly, so we no longer need to fetch the same data from global hashmap

**Comment by hlinnaka** [comment]
> I think this is a correct move overall, but I would rather keep model together with constructor functions.
> 
> Repository provides public `RepositoryTimeline` and `TimelineImpl`, so IMO the way how to transform these structures to present them to outside world is more of a presentation layer task which is represented by models/routes

Hmm, with that thought, I guess the functions belong in routes.rs. And while we're at it, `get_local_timelines` too, because it constructs a list of TimelineInfos. Makes sense; I added a commit to move those, too. WDYT?

**Review by aome510 — APPROVED** [review]

**Comment by LizardWizzard** [comment]
> I guess the functions belong in routes.rs

To me both `models.rs` and `routes.rs` form the presentation layer, so keeping constructor functions along with model adheres to this placement principle. Another way to think about these functions is to treat them as just From impls. In this particular case we cannot use From because we need additional arguments. But if it was possible I would say it would've been an idiomatic way to represent such transformation. And for From it is typical to place it near the struct it is being implemented for

+1 for moving `list_local_timelines` to routes.rs

Also I think with other misc stuff (e g State with corresponding helpers + parameters parsing) sitting in routes.rs it may be a good idea to split them off into some utils.rs

**Comment by hlinnaka** [comment]
> > I guess the functions belong in routes.rs
> 
> To me both `models.rs` and `routes.rs` form the presentation layer, so keeping constructor functions along with model adheres to this placement principle. Another way to think about these functions is to treat them as just From impls. In this particular case we cannot use From because we need additional arguments. But if it was possible I would say it would've been an idiomatic way to represent such transformation. And for From it is typical to place it near the struct it is being implemented for

I'd like to keep `models.rs` as plain data objects, representing just the API. These functions reach pretty deep into the system, so I don't think they belong in `models.rs`. For example, if you pass `include_non_incremental_physical_size=true`, it will call into functions that scan the local directory to calculate the total size. A simple constructor that just stores its arguments in the struct would indeed be better represented by `From` implementations, or functions close to the struct, but these are a lot more complicated.

But as you said, `routes.rs` is part of the presentation layer, so I think that's a suitable place for these.

> Also I think with other misc stuff (e g State with corresponding helpers + parameters parsing) sitting in routes.rs it may be a good idea to split them off into some utils.rs

I won't try to squeeze that into this PR, but yeah, it might make sense. All the handler functions follow the same pattern:

1. parse request, extra query parameters
2. check permissions
3. spawn_blocking(perform the operation)
4. convert result to JsonResponse

**Comment by LizardWizzard** [comment]
> These functions reach pretty deep into the system

I can partially agree with that. Functions use getters available on the passed timeline struct and dont go deeper by themselves. If they are so internal to the system it makes sense to keep the whole thing in timelines.rs as it was before

> but these are a lot more complicated

It is a fair point. However I do not see much difference with them being in routes.rs. This is the same construction logic that now is detached from both, the model it is producing and it still digs into system internals to fetch needed data to present it. From my point of view construction logic is closer to the model rather than to http handling stuff (or even hypotetical utils.rs). To me this is the model that decides how to construct itself from provided timeline. So for me it rather spreads the logic across different places. But I know that this is highly subjective opinion, so I wont insist and if you feel that this is better lets keep it this way :)

> I won't try to squeeze that into this PR

Yeah, agree, this is a separate change

**Pull request merged**